### PR TITLE
fix(ZoneIngress): fix no pointer panic for advertised address resolving

### LINKED ISSUE #10421: ZoneIngress can't start when `networking.advertisedAddress` can't be resolved
### What happened?

https://github.com/kumahq/kuma/blob/444cf2d48bfb8d40301fd9f5e525b3d712143aba/pkg/xds/context/mesh_context_builder.go#L339-L343

when `ResolveZoneIngressPublicAddress` returns error, kuma-cp panics while calling `zi.GetMeta().GetName()`:

> 2024-06-07T11:39:47.570Z	ERROR	xds-server.dataplane-sync-watchdog	OnTick() failed	{"dataplaneKey": {"Mesh":"","Name":"kong-mesh-egress-59788c5d97-p8hqp.kong-mesh-system"}, "error": "runtime error: invalid memory address or nil pointer dereference", "errorVerbose": "runtime error: invalid memory address or nil pointer dereference\ngithub.com/kumahq/kuma/pkg/util/watchdog.(*SimpleWatchdog).onTick.func1\n\tgithub.com/kumahq/kuma@v0.0.0-20240201085611-d176c947ae41/pkg/util/watchdog/watchdog.go:62\nruntime.gopanic\n\truntime/panic.go:914\nruntime.panicmem\n\truntime/panic.go:261\nruntime.sigpanic\n\truntime/signal_unix.go:861\ngithub.com/kumahq/kuma/pkg/core/resources/apis/mesh.(*ZoneIngressResource).GetMeta\n\tgithub.com/kumahq/kuma@v0.0.0-20240201085611-d176c947ae41/pkg/core/resources/apis/mesh/zz_generated.resources.go:2869\ngithub.com/kumahq/kuma/pkg/xds/context.(*meshContextBuilder).fetchResourceList.func1\n\tgithub.com/kumahq/kuma@v0.0.0-20240201085611-d176c947ae41/pkg/xds/context/mesh_context_builder.go:326\ngithub.com/kumahq/kuma/pkg/xds/context.modifyAllEntries\n\tgithub.com/kumahq/kuma@v0.0.0-20240201085611-d176c947ae41/pkg/xds/context/mesh_context_builder.go:356\ngithub.com/kumahq/kuma/pkg/xds/context.(*meshContextBuilder).fetchResourceList\n\tgithub.com/kumahq/kuma@v0.0.0-20240201085611-d176c947ae41/pkg/xds/context/mesh_context_builder.go:312\ngithub.com/kumahq/kuma/pkg/xds/context.(*meshContextBuilder).BuildGlobalContextIfChanged\n\tgithub.com/kumahq/kuma@v0.0.0-20240201085611-d176c947ae41/pkg/xds/context/mesh_context_builder.go:213\ngithub.com/kumahq/kuma/pkg/xds/context.(*meshContextBuilder).BuildIfChanged\n\tgithub.com/kumahq/kuma@v0.0.0-20240201085611-d176c947ae41/pkg/xds/context/mesh_context_builder.go:99\ngithub.com/kumahq/kuma/pkg/xds/cache/mesh.(*Cache).GetMeshContext.func1\n\tgithub.com/kumahq/kuma@v0.0.0-20240201085611-d176c947ae41/pkg/xds/cache/mesh/cache.go:61\ngithub.com/kumahq/kuma/pkg/xds/cache/once.RetrieverFunc.Call\n\tgithub.com/kumahq/kuma@v0.0.0-20240201085611-d176c947ae41/pkg/xds/cache/once/cache.go:43\ngithub.com/kumahq/kuma/pkg/xds/cache/once.(*Cache).GetOrRetrieve.func1\n\tgithub.com/kumahq/kuma@v0.0.0-20240201085611-d176c947ae41/pkg/xds/cache/once/cache.go:67\ngithub.com/kumahq/kuma/pkg/xds/cache/once.(*once).Do.func1\n\tgithub.com/kumahq/kuma@v0.0.0-20240201085611-d176c947ae41/pkg/xds/cache/once/once.go:13\nsync.(*Once).doSlow\n\tsync/once.go:74\nsync.(*Once).Do\n\tsync/once.go:65\ngithub.com/kumahq/kuma/pkg/xds/cache/once.(*once).Do\n\tgithub.com/kumahq/kuma@v0.0.0-20240201085611-d176c947ae41/pkg/xds/cache/once/once.go:12\ngithub.com/kumahq/kuma/pkg/xds/cache/once.(*Cache).GetOrRetrieve\n\tgithub.com/kumahq/kuma@v0.0.0-20240201085611-d176c947ae41/pkg/xds/cache/once/cache.go:59\ngithub.com/kumahq/kuma/pkg/xds/cache/mesh.(*Cache).GetMeshContext\n\tgithub.com/kumahq/kuma@v0.0.0-20240201085611-d176c947ae41/pkg/xds/cache/mesh/cache.go:52\ngithub.com/kumahq/kuma/pkg/xds/context.AggregateMeshContexts\n\tgithub.com/kumahq/kuma@v0.0.0-20240201085611-d176c947ae41/pkg/xds/context/aggregate_mesh_context.go:29\ngithub.com/kumahq/kuma/pkg/xds/sync.(*DataplaneWatchdog).syncEgress\n\tgithub.com/kumahq/kuma@v0.0.0-20240201085611-d176c947ae41/pkg/xds/sync/dataplane_watchdog.go:218\ngithub.com/kumahq/kuma/pkg/xds/sync.(*DataplaneWatchdog).Sync\n\tgithub.com/kumahq/kuma@v0.0.0-20240201085611-d176c947ae41/pkg/xds/sync/dataplane_watchdog.go:83\ngithub.com/kumahq/kuma/pkg/xds/sync.(*dataplaneWatchdogFactory).New.func2\n\tgithub.com/kumahq/kuma@v0.0.0-20240201085611-d176c947ae41/pkg/xds/sync/dataplane_watchdog_factory.go:43\ngithub.com/kumahq/kuma/pkg/util/watchdog.(*SimpleWatchdog).onTick\n\tgithub.com/kumahq/kuma@v0.0.0-20240201085611-d176c947ae41/pkg/util/watchdog/watchdog.go:70\ngithub.com/kumahq/kuma/pkg/util/watchdog.(*SimpleWatchdog).Start\n\tgithub.com/kumahq/kuma@v0.0.0-20240201085611-d176c947ae41/pkg/util/watchdog/watchdog.go:40\nruntime.goexit\n\truntime/asm_amd64.s:1650"}

### PR DESCRIPTION
**Pull request by Icarus9913**
Refer to #10421, we made a mistake in the past to use a no pointer variable which leads to a panic

close #10421

### Checklist prior to review

- [x] [Link to relevant issue][1] as well as docs and UI issues --
- [x] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [ ] Tests (Unit test, E2E tests, manual test on universal and k8s) --
  - Don't forget `ci/` labels to run additional/fewer tests
- [ ] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [ ] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/xds/context/mesh_context_builder.go [hunk 1]
```diff
@@ -341,9 +341,11 @@ func (m *meshContextBuilder) fetchResourceList(ctx context.Context, resType core
 			if !ok {
 				return nil, errors.New("entry is not a zoneIngress this shouldn't happen")
 			}
+
+			zoneIngressName := zi.GetMeta().GetName()
 			zi, err := xds_topology.ResolveZoneIngressPublicAddress(m.ipFunc, zi)
```
Comment (michaelbeaumont) [inline pkg/xds/context/mesh_context_builder.go:346]:
```suggestion
			resolvedZoneIngress, err := xds_topology.ResolveZoneIngressPublicAddress(m.ipFunc, zi)
```

File: pkg/xds/context/mesh_context_builder.go [hunk 2]
```diff
@@ -341,9 +341,11 @@ func (m *meshContextBuilder) fetchResourceList(ctx context.Context, resType core
...
 			if err != nil {
-				l.Error(err, "failed to resolve zoneIngress's domain name, ignoring zoneIngress", "name", zi.GetMeta().GetName())
+				l.Error(err, "failed to resolve zoneIngress's domain name, ignoring zoneIngress", "name", zoneIngressName)
 				return nil, nil
 			}
 			return zi, nil
```
Comment (michaelbeaumont) [inline pkg/xds/context/mesh_context_builder.go:351]:
```suggestion
			return resolvedZoneIngress, nil
```

**Review by michaelbeaumont — APPROVED** [review]
Hmm it took me a while to see why this is ok, I think the problem is that we overwrite a valid object with an invalid object, maybe it should just have a different name

**Review by michaelbeaumont — APPROVED** [review]

**Review by jakubdyszkiewicz — CHANGES_REQUESTED** [review]
Maybe I don't see something, but how does it fix the issue?
If `zi` is nil, wouldn't `ResolveZoneIngressPublicAddress` resolve nil, but also we still have a panic in `zi.GetMeta()`, no?

Can we maybe have a unit test to reproduce it the panic?

**Review by jakubdyszkiewicz — APPROVED** [review]
ok, I see it now

**Pull request merged**